File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        oldToCopy = {}

        curr = head
        while curr:
            copy = Node(curr.val)
            oldToCopy[curr] = copy
            curr = curr.next

        curr = head
        while curr:
            copy = oldToCopy.get(curr)
            copyNext = oldToCopy.get(curr.next, None)
            copy.next = copyNext

            copyRandom = oldToCopy.get(curr.random, None)
            copy.random = copyRandom
            curr = curr.next

        return oldToCopy.get(head, None)

File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.js
/**
 * // Definition for a Node.
 * function Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */
var copyRandomList = function(head) {
    const oldToCopy = new Map();

    let curr = head
    while (curr) {
        const copy = new Node(curr.val);
        oldToCopy.set(curr, copy);
        curr = curr.next
    }

    curr = head
    while (curr) {
        const copy = oldToCopy.get(curr);
        const copyNext = oldToCopy.get(curr.next) || null;
        copy.next = copyNext;

        const copyRandom = oldToCopy.get(curr.random) || null;
        copy.random = copyRandom;
        curr = curr.next;
    }
    return oldToCopy.get(head) || null;
};